include/zephyr/net/tftp.h

```diff
@@ -53,41 +44,133 @@ struct tftpc {
 #define TFTPC_REMOTE_ERROR       -4
 #define TFTPC_RETRIES_EXHAUSTED  -5
 
-/* @brief This function gets "file" from the remote server.
+/**
+ * @brief TFTP Asynchronous Events notified to the application from the module
+ *        through the callback registered by the application.
+ */
+enum tftp_evt_type {
+	/** DATA event when data is received from remote server.
+	 *
+	 * @note DATA event structure contains payload data and size.
+	 */
+	TFTP_EVT_DATA,
+
+	/**  ERROR event when error is received from remote server.
+	 *
+	 * @note ERROR event structure contains error code and message.
+	 */
+	TFTP_EVT_ERROR
+};
+
+/** @brief Parameters for data event. */
+struct tftp_data_param {
+	uint8_t *data_ptr;         /**< Pointer to binary data. */
+	uint32_t len;              /**< Length of binary data. */
+};
+
+/** @brief Parameters for error event. */
+struct tftp_error_param {
+	char *msg;                 /**< Error message. */
+	int code;                  /**< Error code. */
+};
+
+/**
+ * @brief Defines event parameters notified along with asynchronous events
+ *        to the application.
+ */
+union tftp_evt_param {
+	/** Parameters accompanying TFTP_EVT_DATA event. */
+	struct tftp_data_param data;
+
+	/** Parameters accompanying TFTP_EVT_ERROR event. */
+	struct tftp_error_param error;
+};
+
+/** @brief Defines TFTP asynchronous event notified to the application. */
+struct tftp_evt {
+	/** Identifies the event. */
+	enum tftp_evt_type type;
+
+	/** Contains parameters (if any) accompanying the event. */
+	union tftp_evt_param param;
+};
+
+/**
+ * @typedef tftp_callback_t
+ *
+ * @brief Asynchronous event notification callback registered by the
+ *        application.
+ *
+ * @param[in] evt Event description along with result and associated
+ *                parameters (if any).
+ */
+typedef void (*tftp_callback_t)(const struct tftp_evt *evt);
+
+/**
+ * @brief TFTP client definition to maintain information relevant to the
+ *        client.
+ */
+struct tftpc {
+	/** Control Block that represents the remote server */
+	struct sockaddr *server;
+
+	/** Receive buffer used for packet in RX path. */
+	uint8_t rx_buf[TFTPC_MAX_BUF_SIZE];
+
+	/** Transmit buffer used for packet in TX path. */
+	uint8_t tx_buf[TFTPC_MAX_BUF_SIZE];
+
+	/** Event notification callback */
+	tftp_callback_t callback;
+
+	/** Internal. Socket handle. */
+	int sock;
```

Comment:
(looks like my comment was gone)
Here we use client.sock to provide exclusive access to GET or PUT transfer, preventing re-entry from the same client from the same or different thread. We can save the usage of mutex as well. Also there is no requirement for simultaneous GET and PUT transfer support, so one socket handle is good enough.

Reply:
This kind of protection is not ideal, and susceptible of race conditions. There is a small window between `client->sock` check, and the actual socket creation, where the protection would fail, allowing to manipulate client context from two different threads.

Reply:
You're right. Let's remove the protection by socket handle check, and use two semaphore to protect GET and PUT respectively.